### PR TITLE
[WIP] Modular Hardsuits | Iron Man Edition.

### PR DESCRIPTION
**Pull request by Bibby0110**
# Github documenting your Pull Request

Basically turns all hardsuits on the station into modular ones. What does this mean you ask? It allows you to install various upgrades into the hardsuit. Stuff like jetpacks, tools, shock resistance, oxygen recyclers, assisted exoskeletons and so on.

Also adds a new Roboticist gadget, which is practically just an iron man suit (Sprites stolen from the old flight suit) that allows the mounting of lethal and non-lethal weapons. (Disabler, stun baton hands, etc) Lethal weapons are ordered through cargo, (Basically a worse WT rifle that is stuck to your arm, has less damage, firerate and ammo)

The 'Iron Man'/ACE (Advanced Combat Exoskeleton) suit allows the mounting of ALL upgrades.

Security can mount defensive upgrades + non-lethal non-ranged weapons.

Engineering suits get a reduction in power consumption.

Medical suits can mount medically relevant upgrades, (Integrated defib and medical dispenser)

Regular EVA suits/soft suits cannot mount upgrades  

Nukeops can purchase one and an upgrade bundle giving access to better weapons (Shoulder mounted mini-rockets anyone?)

Suits now require power cells to be inserted for basic functionality. You'll be able to not use a cell but the A/C in the suit won't work meaning you'll be cooked alive slowly. You will also walk slower. There's an emergency battery built into all the suits so even if you run out you'll have a brief window to slot in a new cell or find one.

Everything is controlled through an integrated TGUI showing suit stats (Power capacity, charge %, power generation, installed modules and so on)

- [ ] Basic framework
- [ ] Power cell mounting
- [ ] Basic life support and cell integration
- [ ] Modules framework
- [ ] Utility modules (ARC reactor, oxygen recycler, welding googles, pAI integration(?) etc)
- [ ] Mobility modules (Jetpack, flight mode, powered exoskeleton)
- [ ] Defensive modules (Extra armor, EMP resistance, etc)
- [ ] Offensive modules (Stun hands, disabler arms, lethal weapons)
- [ ] Station variants (Engineering suits primarily)
- [ ] Syndicate variant
- [ ] Cargo crates
- [ ] Research integration (RnD + mech assembler)
- [ ] Github Changelog

<details>
<summary>Planned modules</summary>

**Utility Modules**
- Extra battery slot. (Allows 2 mounted battery cells)
- Photovoltaic generator (Minor passive power generation while in space)
- Experimental ARC reactor (Pending name, less advanced power generator)
- ARC reactor (Pending name, highly advanced passive power generator)
- Efficiency module (Less power consumption)
- Welding googles. (Helmet functions like a welding helmet. No sight reduction, just like current hardsuits)
- pAI integration. (pAI gets to control suit when there's no occupant, or manage internal systems while occupant is inside.) (Experimental, unlikely to be done)
- Oxygen recycler. (Reduces oxygen consumption)
- Emergency Medical Dispenser. (Dispenses cocktail of stabilizing chems when getting close to crit. Will not heal out of crit.)
- Emergency Defibrillator (Defib running on internal power for those space medics. Goes through hardsuits. High delay and power consumption)
- Pneumatic Injector (Goes through suits, injects any drugs you put in it.)

**Mobility Modules**

- Jetpack. (Runs on your oxygen tank like normal.)
- Flight mode. (Copy of the flight suit, you can fly indoors, you get knocked down when hitting someone, potentially strips you of the suit when crashing into walls.)
- Powered exoskeleton (Less slowdown.)

**Defensive Modules**
(Max 1 armor upgrade)
- Structural Reinforcement (Extra armor, slowdown)
- Deflector Array (Extra armor, uses power, no slowdown.)
- Localized Deflector Array (Extra armor against specific damage type. Uses power. No slowdown)
(No restriction)
- EMP resistance. (Internal cells are drained less when EMP'd.)
- Atmospheric Reinforcement (Upgrades to atmos suit level of atmos resistance)
- Radiation Shielding (Upgrades to atmos suit level of rad resistance)

**Offensive Modules**

- Stun hands. (Stunbaton gloves but they're retractable and no-drop. High power consumption)
- Disabler. (Summons disabler at will. Runs on integrated power cell.)
- Rifle (Summons a worse WT rifle. Automagically prints ammo when retracted, high power consumption)
Syndicate only:
- Mounted C20-r (Regular C20-r, fast ammo printing when retracted, reduced power consumption)
- M-546P (Mounted minigun, low initial ammo capacity, gradually creates more. Slows down the suit)
- Shoulder Mounted mini-RPG (Fires worse rockets, short stun time, medium damage. Reloads manually from a rocket pod item)
- Integrated E-swords. (No-drop retractable e-swords, nothing special.)

**Suit specific**
Engineering suits get appropriate welding upgrades and atmospheric upgrades. All have reduced power consumption through hidden upgrade.
Regular EVA soft-suits do not take upgrades but still require power cells.

</details>

# Wiki Documentation

# Changelog

:cl:  
rscadd: Modular Hardsuits
/:cl:

### CODE REVIEW

**Review by alexkar598 — CHANGES_REQUESTED** [review]
code will not compile

**Comment by boodaliboo** [comment]
based gamer

**Comment by Bibby0110** [comment]
> code will not compile

Code very much compiles right now

File: README.md [hunk 1]
```diff
@@ -35,7 +35,7 @@ code tab of https://github.com/yogstation13/Yogstation
 (note: this will use a lot of bandwidth if you wish to update and is a lot of
 hassle if you want to make any changes at all, so it's not recommended.)
 
-## The Yogstation codebase recommends compiling using version [514.1554](www.byond.com/download/build/514/514.1554_byond.exe) and may potentially NOT work on newer or older versions.
+## The Yogstation codebase recommends compiling using version [514.1554](https://www.byond.com/download/build/514/514.1554_byond.exe) and may potentially NOT work on newer or older versions.
```
Comment (adamsong) [inline README.md:38]: Undocumented change

**Review by adamsong — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (Bibby0110) [reply README.md:38]: go away

**Comment by Bluedazzled** [comment]
hmmm
image
sure does feel familiar.

**Comment by Bibby0110** [comment]
> hmmm
> image
> sure does feel familiar.

Probably cause I stole the basic idea from there.

File: code/modules/clothing/spacesuits/modular_suits/_module.dm [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+/obj/item/hardsuit_upgrade/proc/removal_effect(suit)
+	return
+
+/obj/item/hardsuit_upgrade/proc/can_toggle()
+	return TRUE
+
+/obj/item/hardsuit_upgrade/proc/toggle(suit)
+	if(!can_toggle())
+		return
+	enabled = !enable
```
Comment (Bibby0110) [inline code/modules/clothing/spacesuits/modular_suits/_module.dm]:
```suggestion
	enabled = !enabled
```

File: code/modules/clothing/spacesuits/modular_suits/_modular_suits.dm [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/datum/modular_hardsuit
+	var/inserted_modules
+	var/obj/item/clothing/suit/space/hardsuit/suit
+
+/datum/modular_hardsuit/New(new_suit)
+	if(!new_suit || !istype(new_suit(/obj/item/clothing/suit/space/hardsuit)))
```
Comment (Bibby0110) [inline code/modules/clothing/spacesuits/modular_suits/_modular_suits.dm]:
```suggestion
	if(!new_suit || !istype(new_suit, /obj/item/clothing/suit/space/hardsuit))
```

File: code/modules/clothing/spacesuits/modular_suits/_modular_suits.dm [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		qdel(src)
+		stack_trace("Modular hardsuit datum initialized with no suit attached!")
+		return
+	suit = new_suit
+	inserted_modules = list()
+	if(suit.initial_modules.length)
```
Comment (Bibby0110) [inline code/modules/clothing/spacesuits/modular_suits/_modular_suits.dm]:
```suggestion
	if(length(suit.initial_modules))
```

File: code/modules/clothing/spacesuits/hardsuit.dm [hunk 5]
```diff
@@ -103,7 +103,25 @@
 	var/helmettype = /obj/item/clothing/head/helmet/space/hardsuit
 	var/obj/item/tank/jetpack/suit/jetpack = null
 
+	//Modular start
+	var/allowed_modules
+
+	var/initial_modules = list()
+
+	var/datum/modular_hardsuit/mod_suit
```
Comment (alexkar598) [inline code/modules/clothing/spacesuits/hardsuit.dm:111]: What's the point of using a datum here, just have everything on the suit itself

Comment (Bibby0110) [reply code/modules/clothing/spacesuits/hardsuit.dm:111]: you're disgusting

Comment (Bibby0110) [reply code/modules/clothing/spacesuits/hardsuit.dm:111]: disgusting but correct

**Comment by Bibby0110** [comment]
I need to figure out a different way to do this

**Pull request closed without merging**